### ISSUE TITLE
Support for bulk operations and queries (solving the N+1 problem)

### ISSUE THREAD

**Issue by ahmednfwela** [zitadel/zitadel#6186]
### Preflight Checklist

- [X] I could not find a solution in the existing issues, docs, nor discussions
- [X] I have joined the [ZITADEL chat](https://zitadel.com/chat)

### Describe your problem

we require those operations in order to avoid the N+1 problem when scaling our userbase:
1. get multiple users by ids
`where id IN [ids]`
2. get multiple users by metadata field
`where metadata["legacyId"] IN [legacyIds]`
3. get multiple users by ids or metadata field (in the same request)
`where id IN [ids] OR metadata["legacyId"] IN [ids]`
4. get metadata of multiple users at once
or preferably include the metadata in the `User` object from the API via a `IncludeMetada` flag in the request
5. get roles of multiple users at once for all/specific projects
by default it should selects all projects that are visible to the service account
6. (optional if the first 5 are implemented) List all users in an organization, for the ability to sync with meilisearch/algolia instance
7. (optional if the first 5 are implemented) have an action that is able to sync user updates with meilisearch/algolia instance
8. Bulk import users and include the metadata in the import request

### Describe your ideal solution

As a developer, I want to be able to query a lot of users fast, instead of the current API which requires us to send a request for every user, since it doesn't scale in terms of cost and performance.

### Version

v2.30.0

### Environment

ZITADEL Cloud

**Comment by hifabienne** [comment]
Hei @ahmednfwela 
Thanks for creating the improvement issue.
I think everything you described is a valid point and I can see the use case for it.
We do not yet have a timeline on when to implement it, so I will put it into the product backlog for now.
If you need it more urgent we also appreciate and accept PR, if you are interested in contributing some of the functionality.

For the Bulk import there is already another issue: https://github.com/zitadel/zitadel/issues/5524

**Comment by ahmednfwela** [comment]
@hifabienne I will be creating some small PRs to address these points one by one, I appreciate your assistance, Thanks!

**Comment by hifabienne** [comment]
Hei @ahmednfwela Thank you very much. We really appreciate your contribution

**Comment by lukaslihotzki-f** [comment]
This issue currently mentions many specific bulk queries. Alternatively, GraphQL support could be added in Zitadel so bulk queries (including those explained in the description) can be specified generically. There is an existing [Go GraphQL package](https://github.com/graphql-go/graphql). Would this solution be acceptable too?

**Comment by ahmednfwela** [comment]
I am not familiar with graphql and I am not sure how will it integrate with the grpc model.

but if there is a better solution , I am up for it

**Comment by hifabienne** [comment]
I think it doesn't really matter if it is GraphQL or not, as graphQL isn't really about bulk operations, but about generic input/output.
For the moment i think we will stick to grpc and rest, as GraphQL would add a lot of complexity to maintain.
Of course there would be the possibility to make a separat GraphQL project which mappes the requests to the current API.

**Comment by muhlemmer** [comment]
In order not to overload review capacity, we should consider each and very point of the original proposal separately, before a PR is send. For example https://github.com/zitadel/zitadel/pull/6361 review process took too long IMO, because there was no clear consensus on the API change prior to implementation

## Filtering on metadata

> get multiple users by metadata field
where metadata["legacyId"] IN [legacyIds]

Concerns me. We don't have an index on metadata keys. We could add one, but that has a price too of longer write locks in CockroachDB. This would create issues where an action would update metadata (which pushes to the eventstore) and we use `Trigger` to update the projections before we can continue serving the request. At the moment Triggers are the most expensive components in zitadel when measuring request times.

I'm not saying it's bad, but we have to consider if it's worth paying the performance price if only 1 of few people need it. Can't your problem not be solved otherwise?

For example, why metadata with a legacyID if you can create users with [your own ID ](https://zitadel.com/docs/apis/resources/user_service/user-service-add-human-user) already?

```
Request Body required
|- userId optionally set your own id unique for the user
|- Possible values: non-empty and <= 200 characters
```

## Bulk

> 6. (optional if the first 5 are implemented) List all users in an organization, for the ability to sync with meilisearch/algolia instance
>7. ....
> 8. Bulk import users and include the metadata/roles in the import request

Dealing with large chunks of data in Unary grpc calls is painful, as the whole result set needs to be allocated in memory, which then has to be garbage collected after completion. I've opened https://github.com/zitadel/zitadel/issues/6686 for the v2 API to propose a streaming API for the import and export purposes.

We already have some [import and export](https://zitadel.com/docs/category/apis/resources/admin/import-export) functionality , but that uses complete chunks and the API isn't the most future proof.

## Metadata on user object

> or preferably include the metadata in the User object from the API via a IncludeMetada flag in the request

yes, completely agree. As a matter of fact I'm currently optimizing the introspection and token endpoints and one thing that bugged me is all the separate queries that are happening. First there's a `GetUserByID`, then there is a `GetUserMetadata`. I'm in the process of rewriting that in order to resolve all user related data in a single query. This change would benefit towards the usecases in this PR as well, I hope.

**Comment by ahmednfwela** [comment]
> I've opened https://github.com/zitadel/zitadel/issues/6686 for the v2 API to propose a streaming API for the import and export purposes.

what we need isn't just an initial export/import of the data, we also need a live stream of users that have been added/changed/deleted in order to do live sync with the database.

if that is available, then we don't need to mess with zitadel's API, since we can just use our own indexing solution.

**Comment by mogita** [comment]
Adding a use case that could largely benefit from the proposal of 1) including metadata in the users search or users grants search endpoints; 2) making queries on user metadata to fetch a list of users.

We are using Zitadel to host our multi-tenancy infrastructure. Our customers can import their contractors into our system and we hold these contractors in Zitadel. The customers are asking us if we can provide a source where the contractors were imported from, because we support multiple vendors from where our customers can import the contractors, so our customers can distinguish which contractor comes from which vendor.

User metadata is an obvious place to store the source info even source id, etc. However lacking metadata in Users list (we've experimented with fetching by a Service User via `/users/grants/_search` and `/users/_search`) creates a major blocker when we show a list of contractors to our customers. To find out the source of the contractors, they'll need to view contractors one by one.

Currently we're depending on another micro-service to store the metadata and attach the info to the list when queried. This workaround doesn't hold strong for the long run and it'll be really nice to solve it totally within Zitadel.
